pkg/operator/target_status.go

```diff
@@ -0,0 +1,936 @@
+	if err := registry.Register(targetStatusDuration); err != nil {
+		return nil, err
+	}
+
+	ch := make(chan event.GenericEvent, 1)
+
+	dataManager, err := builder.build()
+	if err != nil {
+		return nil, err
+	}
+
+	reconciler := &targetStatusReconciler{
+		ch:      ch,
+		opts:    opts,
+		manager: dataManager,
+		logger:  logger,
+	}
+
+	return reconciler, nil
+}
+
+type emptyObject struct{}
+
+func (r *targetStatusReconciler) Start(ctx context.Context, tickerBuilder func() Ticker) error {
+	r.tickerBuilder = tickerBuilder
+	r.queue()
+	return nil
+}
+
+// Queues a reconcile.
+func (r *targetStatusReconciler) queue() {
+	r.ch <- event.GenericEvent{
+		Object: &appsv1.DaemonSet{},
+	}
+}
+
+// Reconcile polls the collector pods, fetches and aggregates target status and upserts
+// into each PodMonitoring's Status field.
+func (r *targetStatusReconciler) Reconcile(ctx context.Context, _ reconcile.Request) (reconcile.Result, error) {
+	// minDelay := time.NewTimer(10 * time.Second)
+
+	minDelay := r.tickerBuilder()
+	// defer minDelay.Stop()
+
+	onceLog := new(sync.Once)
+
+	// Wrapper function so we could exit early.
+	err := func() error {
+		now := time.Now()
+		if err := poll(ctx, r.logger, r.opts, r.manager); err != nil {
+			// Not having data is not an error.
+			if apierrors.IsNotFound(err) {
+				onceLog.Do(func() {
+					r.logger.Error(err, "Kubernetes config not found")
+				})
+				return nil
+			}
+			return err
+		}
+		// In case Kubernetes config is removed and re-added.
+		onceLog = new(sync.Once)
+
+		duration := time.Since(now)
+		targetStatusDuration.WithLabelValues().Set(float64(duration.Milliseconds()))
+		return nil
+	}()
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Check if we beat the timer, otherwise wait.
+	<-minDelay.Channel()
+	// Re-trigger event
+	r.queue()
+
+	return reconcile.Result{}, nil
+}
+
+func poll(ctx context.Context, logger logr.Logger, opts Options, manager dataManager) error {
+	targets, err := fetchTargets(ctx, logger, opts, manager)
+	if err != nil {
+		return err
+	}
+
+	return populateTargets(ctx, logger, manager, targets)
+}
+
+func fetchTargets(ctx context.Context, logger logr.Logger, opts Options, manager dataManager) ([]*prometheusTargetsResultAndTime, error) {
+	podMap, err := manager.getPrometheusPods(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Set up pod job queue and jobs
+	podDiscoveryChan := make(chan prometheusPod, opts.TargetPollConcurrency)
+	wg := sync.WaitGroup{}
+	wg.Add(int(opts.TargetPollConcurrency))
+
+	// Must be unbounded or else we deadlock.
+	targetChan := make(chan *prometheusTargetsResultAndTime)
+	for i := uint16(0); i < opts.TargetPollConcurrency; i++ {
+		// Wrapper function to handle errors.
+		go func(thread uint16) {
+			fetcher, err := manager.getTargetFetcher(ctx, thread)
+			if err != nil {
```

Comment:
Also, as mentioned before, if we can get rid of the `dataManager` interface, this flow can hopefully just become a single call, i.e.
```go
for i := 0; i < opts.TargetPollConcurrency; i++ {
	go func {
		defer wg.Done()
		// Process incoming pods and send results to target channel, blocking until pod channel closes.
		processTargetQueue(ctx, logger, podCh, targetCh)
	}
}
```

Reply:
I ended up turning it into a builder on the outside of the concurrency loop and use the built object inside because the object has some common state. I specifically test that concurrency is working in the unit test. Let me know what you think of this interface in the new iteration of the code.